File: code/_ROS_Generator/Generator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.RegularExpressions;
using System.Windows;
using System.Linq;

namespace ROS_Generator
{
    // Extending CMakeLists template class
    public partial class CMakeLists
    {
        public string project_name;
        public List<string> messages;
        public List<string> services;

        public CMakeLists()
        {
            project_name = "";
            messages = new List<string>();
            services = new List<string>();
        }
    }

    // Extending Package.xml template class
    public partial class package_xml
    {
        public string project_name;

        public package_xml()
        {
            project_name = "";
        }
    }

    // Extending Component base class template
    public partial class Component_Base_cpp
    {
        public string project_name;

        public Component_Base_cpp()
        {
            project_name = "";
        }
    }

    // Extending Component hpp template
    public partial class Component_hpp
    {
        public string define_guard;
        public string project_name;
        public string node_name;
        public string comp_name;
        public List<Visitor.Publisher> publishers;
        public List<Visitor.Subscriber> subscribers;
        public List<Visitor.Service> provided_services;
        public List<Visitor.Service> required_services;
        public List<Visitor.Timer> timers;
        public List<String> topics;
        public List<String> services;

        public Component_hpp()
        {
            node_name = "";
            comp_name = "";
            publishers = new List<Visitor.Publisher>();
            subscribers = new List<Visitor.Subscriber>();
            provided_services = new List<Visitor.Service>();
            required_services = new List<Visitor.Service>();
            timers = new List<Visitor.Timer>();
            topics = new List<String>();
            services = new List<String>();
        }
    }

    class Generator
    {

        public void generate(string input_directory, List<ROS_Package> ros_packages)
        {
            foreach (var package in ros_packages)
            {
                string package_dir = input_directory + package.name;

                string package_ws = package_dir + "\\" + package.name + "_ws";
                string package_src = package_ws + "\\" + "src";
                string package_main = package_src + "\\" + package.name;

                // Create Workspace directory
                Directory.CreateDirectory(package_ws);

                // Create src directory
                Directory.CreateDirectory(package_src);

                dot_catkin_ws new_ws = new dot_catkin_ws();
                var generated_text = new_ws.TransformText();
                System.IO.File.WriteAllText(package_ws + "\\.catkin_workspace", generated_text);

                // Create the package directory
                Directory.CreateDirectory(package_main);

                // Create the include directory
                Directory.CreateDirectory(package_main + "\\include");

                // Create the src directory
                Directory.CreateDirectory(package_main + "\\src");

                // Create msg directory IF necessary
                if (package.messages.Count > 0)
                {
                    Directory.CreateDirectory(package_main + "\\msg");

                    foreach(var msg in package.messages) 
                    {
                        // Create the message files
                        string filename = msg.name;

                        using (StreamWriter fw = new StreamWriter(package_main + "\\msg\\" + filename, false))
                        {
                            fw.Write(msg.definition);
                        }
                    }
                }

                // Create service directory IF necessary
                if (package.services.Count > 0)
                {
                    Directory.CreateDirectory(package_main + "\\srv");

                    foreach (var srv in package.services)
                    {
                        // Create the service files
                        string filename = srv.name;

                        using (StreamWriter fw = new StreamWriter(package_main + "\\srv\\" + filename, false))
                        {
                            fw.Write(srv.definition);
                        }
                    }
                }

                // Create base Component.cpp
                Component_Base_cpp base_comp = new Component_Base_cpp();
                base_comp.project_name = package.name;
                generated_text = base_comp.TransformText();
                System.IO.File.WriteAllText(package_main + "\\src\\Component.cpp", generated_text);

                // Create base Component.h
                Component_Base_hpp base_comp_hpp = new Component_Base_hpp();
                generated_text = base_comp_hpp.TransformText();
                System.IO.File.WriteAllText(package_main + "\\include\\Component.hpp", generated_text);

                // Create CMakeLists file
                CMakeLists cmake_lists = new CMakeLists();
                cmake_lists.project_name = package.name;
                // Setup all messages
                foreach(var msg in package.messages) 
                {
                    cmake_lists.messages.Add(msg.name);
                }
                // Setup all services
                foreach (var srv in package.services)
                {
                    cmake_lists.services.Add(srv.name);
                }
                generated_text = cmake_lists.TransformText();
                System.IO.File.WriteAllText(package_main + "\\CMakeLists.txt", generated_text);   
             
                // Create package.xml file
                package_xml new_package = new package_xml();
                new_package.project_name = package.name;
                generated_text = new_package.TransformText();
                System.IO.File.WriteAllText(package_main + "\\package.xml", generated_text);


                // Generating header file and cpp file for each component
                foreach (var node in package.nodes)
                {
                    foreach (var component in node.components)
                    {
                        // HPP GENERATION
                        Component_hpp new_hpp = new Component_hpp();
                        new_hpp.project_name = package.name;

                        new_hpp.define_guard = component.name.ToUpper();

                        char[] node_txt = { '.', 't', 'x', 't' };
                        new_hpp.node_name = node.name.TrimEnd(node_txt);
                        new_hpp.comp_name = component.name;

                        if (component.timers.Count > 0)
                        {
                            foreach (var timer in component.timers)
                            {
                                new_hpp.timers.Add(timer);
                            }
                        }

                        if (component.subscribers.Count > 0)
                        {
                            foreach (var subscriber in component.subscribers)
                            {
                                char[] msg = {'.', 'm', 's', 'g'};
                                string topic_name = subscriber.topic.TrimEnd(msg);
                                new_hpp.topics.Add(topic_name);
                                new_hpp.subscribers.Add(new Visitor.Subscriber { name = subscriber.name, topic = topic_name });
                            }
                        }

                        if (component.publishers.Count > 0)
                        {
                            foreach (var publisher in component.publishers)
                            {
                                char[] msg = { '.', 'm', 's', 'g' };
                                string topic_name = publisher.topic.TrimEnd(msg);
                                new_hpp.topics.Add(topic_name);
                                new_hpp.publishers.Add(new Visitor.Publisher { name = publisher.name, topic = topic_name});
                            }
                        }

                        // Find unique topics
                        new_hpp.topics = new_hpp.topics.Select(x => x).Distinct().ToList();
                        //Notes.Select(x => x.Author).Distinct();

                        if (component.provided_services.Count > 0)
                        {
                            foreach (var provided in component.provided_services)
                            {
                                char[] srv = { '.', 's', 'r', 'v' };
                                string srv_name = provided.name.TrimEnd(srv);
                                new_hpp.services.Add(srv_name);
                                new_hpp.provided_services.Add(provided);
                            }
                        }

                        if (component.required_services.Count > 0)
                        {
                            foreach (var required in component.required_services)
                            {
                                char[] srv = { '.', 's', 'r', 'v' };
                                string srv_name = required.name.TrimEnd(srv);
                                new_hpp.services.Add(srv_name);
                                new_hpp.required_services.Add(required);
                            }
                        }

                        // Find unique services
                        new_hpp.services = new_hpp.services.Select(x => x).Distinct().ToList();

                        generated_text = new_hpp.TransformText();
                        System.IO.File.WriteAllText(package_main + "\\include\\" + component.name + ".hpp", generated_text);

                        // CPP

                    }

                }

            }

        }
    }
}

File: release/rosmod_v0.1/generator_v1/python/generator.py
# Generator
#
# Generates the ROS Workspace
#
# Author: Pranav Srinivas Kumar
# Date: 2014.12.16

import sys
import os
from Cheetah.Template import Template
import inspect
import collections
# Template Compile step -- Compiling tmpl files in templates
# Generates equivalent .py files
generator_dir = os.path.dirname(os.path.realpath(__file__))
template_dir = os.path.join(generator_dir + "/templates")

# Recursively compile on template files in templates directory
os.system("/usr/local/bin/cheetah compile " + template_dir + "/*.tmpl")
ros_templates = os.path.realpath(os.path.abspath(os.path.join(os.path.split(inspect.getfile(inspect.currentframe()))[0], "templates")))
if ros_templates not in sys.path:
    sys.path.insert(0, ros_templates)
from package_xml import *
from base_component_hpp import *
from base_component_cpp import *
from msg import *
from srv import *
from component_hpp import *
from component_cpp import *
from nodeMain import *
from CMakeLists import *

from OrderedSet import *

class Generator:
    
    def generate(self, listener, path):

        print "\nGenerating ROS Workspace..."
        # Make the workspace directory
        self.workspace_dir = path + "/" + listener.workspace.name
        print "Workspace Path:", self.workspace_dir
        if not os.path.exists(self.workspace_dir):
            os.makedirs(self.workspace_dir)

        # Generate the .catkin_ws file in the workspace directory
        catkin_ws = "# This file currently only serves to mark the location of a catkin workspace for tool integration "
        filename = ".catkin_workspace"
        with open(os.path.join(self.workspace_dir, filename), 'w') as temp_file:
            temp_file.write(catkin_ws)
            temp_file.close()

        # Make the src directory
        self.src_path = self.workspace_dir + "/src"
        if not os.path.exists(self.src_path):
            os.makedirs(self.src_path)

        # For each package in the ros model
        for package in listener.workspace.packages:
            # Create the package directory
            self.package_path = self.src_path + "/" + package.name
            print package.name, "Path:", self.package_path
            if not os.path.exists(self.package_path):
                os.makedirs(self.package_path)

            # Create the include directory
            self.include = self.package_path + "/include"
            if not os.path.exists(self.include):
                os.makedirs(self.include)
                
            # Create the src directory
            self.src = self.package_path + "/src"
            if not os.path.exists(self.src):
                os.makedirs(self.src)

            # Create the msg directory if there are messages
            if (len(package.messages) > 0):
                self.msg = self.package_path + "/msg"
                if not os.path.exists(self.msg):
                    os.makedirs(self.msg)

            # Create the srv directory if there are services
            if (len(package.services) > 0):
                self.srv = self.package_path + "/srv"
                if not os.path.exists(self.srv):
                    os.makedirs(self.srv)

            # Create package.xml
            package_xml_namespace = {'package_name': package.name}
            t = package_xml(searchList=[package_xml_namespace])
            self.package_xml = str(t)
            with open(os.path.join(self.package_path, "package.xml"), 'w') as temp_file:
                temp_file.write(self.package_xml)

            # Create Component.cpp and Component.hpp
            self.cpp = self.src + "/" + package.name
            self.hpp = self.include + "/" + package.name

            if not os.path.exists(self.cpp):
                os.makedirs(self.cpp)
            base_cpp_namespace = {'hash_include': "#include", 
                                      'package_name': package.name}
            # Populate Base Component cpp template
            t = base_component_cpp(searchList=[base_cpp_namespace])
            self.base_cpp = str(t)
            # Write Component.cpp
            with open(os.path.join(self.cpp, "Component.cpp"), 'w') as temp_file:
                temp_file.write(self.base_cpp)

            if not os.path.exists(self.hpp):
                os.makedirs(self.hpp)
            base_hpp_namespace = {'hash_include': '#include'}
            # Populate Base Component hpp template
            t = base_component_hpp(searchList=[base_hpp_namespace])
            self.base_hpp = str(t)
            # Write Component.hpp
            with open(os.path.join(self.hpp, "Component.hpp"), 'w') as temp_file:
                temp_file.write(self.base_hpp)

            # Create all package messages in msg folder
            for message in package.messages:
                msg_namespace = {'fields': message.fields}
                msg_filename = message.name + ".msg"
                t = msg(searchList=[msg_namespace])
                self.msg_fields = str(t)
                # Write msg file
                with open(os.path.join(self.msg, msg_filename), 'w') as temp_file:
                    temp_file.write(self.msg_fields)

            # Create all package services in srv folder
            for service in package.services:
                srv_filename = service.name + ".srv"
                srv_namespace = {'request_fields': service.request_fields, 
                                 'response_fields': service.response_fields}
                t = srv(searchList=[srv_namespace])
                self.srv_fields = str(t)
                # Write the srv file
                with open(os.path.join(self.srv, srv_filename), 'w') as temp_file:
                    temp_file.write(self.srv_fields)


            # Create a .hpp and .cpp per component definition in package
            for component in package.components:
                component_name  = component.name
                define_guard = component_name.upper()
                topics = []
                for publisher in component.publishers:
                    topics.append(publisher.topic)
                for subscriber in component.subscribers:
                    topics.append(subscriber.topic)
                topics = OrderedSet(topics)
                services = []
                for provided in component.provided_services:
                    services.append(provided)
                for required in component.required_services:
                    services.append(required)
                services = OrderedSet(services)
                hash_include = "#include"
                component_namespace = {'define_guard': define_guard, 
                                       'hash_include': hash_include, 
                                       'package_name': package.name, 
                                       'topics': topics, 
                                       'services': services, 
                                       'component_name': component.name,
                                       'init_business_logic': component.init_business_logic,
                                       'user_includes': component.user_includes,
                                       'user_globals': component.user_globals,
                                       'hpp_globals': component.hpp_globals,
                                       'user_private_variables': component.user_private_variables,
                                       'publishers': component.publishers, 
                                       'subscribers': component.subscribers, 
                                       'provided_services': component.provided_services, 
                                       'required_services': component.required_services, 
                                       'timers': component.timers}
                t = component_hpp(searchList=[component_namespace])
                self.component_hpp_str = str(t)
                # Write the component hpp file
                hpp_filename = component_name + ".hpp"

                with open(os.path.join(self.hpp, hpp_filename), 'w') as temp_file_h:
                    temp_file_h.write(self.component_hpp_str)

                # GENERATING CPP FILES

                t = component_cpp(searchList=[component_namespace])
                self.component_cpp_str = str(t)
                # Write the component cpp file
                cpp_filename = component_name + ".cpp"

                with open(os.path.join(self.cpp, cpp_filename), 'w') as temp_file:
                    temp_file.write(self.component_cpp_str)

            for node in package.nodes:
                node_name = node.name
                hash_include = "#include"
                node_namespace = {'hash_include': hash_include, 
                                  'package_name': package.name,
                                  'node_name': node_name, 
                                  'components': node.components}
                t = nodeMain(searchList=[node_namespace])
                self.nodeMain_str = str(t)
                # Write node main.cpp file
                node_filename = node_name + "_main.cpp"

                with open(os.path.join(self.cpp, node_filename), 'w') as temp_file_nm:
                    temp_file_nm.write(self.nodeMain_str)

            cmake_lists_namespace = {'package_name': package.name,
                                     'packages': package.cmakelists_packages,
                                     'messages': package.messages, 
                                     'services': package.services, 
                                     'catkin_INCLUDE_DIRS': "${catkin_INCLUDE_DIRS}",
                                     'PROJECT_NAME': "${PROJECT_NAME}",
                                     'catkin_LIBRARIES': '${catkin_LIBRARIES}',
                                     'CATKIN_PACKAGE_BIN_DESTINATION':
                                           "${CATKIN_PACKAGE_BIN_DESTINATION}",
                                     'CATKIN_PACKAGE_LIB_DESTINATION': 
                                           "${CATKIN_PACKAGE_LIB_DESTINATION}",
                                     'CATKIN_PACKAGE_INCLUDE_DESTINATION':
                                           "${CATKIN_PACKAGE_INCLUDE_DESTINATION}",
                                     'CATKIN_PACKAGE_SHARE_DESTINATION':
                                           "${CATKIN_PACKAGE_SHARE_DESTINATION}",
                                     'nodes': package.nodes}
            t = CMakeLists(searchList=[cmake_lists_namespace])
            self.cmake_lists = str(t)
            # Write CMakeLists file
            with open(os.path.join(self.package_path, "CMakeLists.txt"), 'w') as temp_file:
                temp_file.write(self.cmake_lists)
            
                

